### ISSUE TITLE
TestRunnerViewPart: SWT Resource was not properly disposed

### ISSUE THREAD

**Issue by jukzi** [eclipse-platform/eclipse.platform#528]
```
java.lang.Error: SWT Resource was not properly disposed
	at org.eclipse.swt.graphics.Resource.initNonDisposeTracking(Resource.java:172)
	at org.eclipse.swt.graphics.Resource.<init>(Resource.java:120)
	at org.eclipse.swt.graphics.Image.<init>(Image.java:668)
	at org.eclipse.jface.resource.URLImageDescriptor.createImage(URLImageDescriptor.java:300)
	at org.eclipse.jface.resource.ImageDescriptor.createResource(ImageDescriptor.java:230)
	at org.eclipse.jface.resource.DeviceResourceManager.allocate(DeviceResourceManager.java:55)
	at org.eclipse.jface.resource.AbstractResourceManager.create(AbstractResourceManager.java:88)
	at org.eclipse.jface.resource.LazyResourceManager.create(LazyResourceManager.java:103)
	at org.eclipse.jface.resource.LocalResourceManager.allocate(LocalResourceManager.java:72)
	at org.eclipse.jface.resource.AbstractResourceManager.create(AbstractResourceManager.java:88)
	at org.eclipse.jface.resource.ResourceManager.createImageWithDefault(ResourceManager.java:195)
	at org.eclipse.jface.action.ActionContributionItem.updateImages(ActionContributionItem.java:1016)
	at org.eclipse.jface.action.ActionContributionItem.updateToolItem(ActionContributionItem.java:776)
	at org.eclipse.jface.action.ActionContributionItem.update(ActionContributionItem.java:739)
	at org.eclipse.jface.action.ActionContributionItem.fill(ActionContributionItem.java:350)
	at org.eclipse.jface.action.ToolBarManager.update(ToolBarManager.java:358)
	at org.eclipse.ui.internal.e4.compatibility.ActionBars.updateActionBars(ActionBars.java:68)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.configureToolBar(TestRunnerViewPart.java:2129)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.createPartControl(TestRunnerViewPart.java:1893)
...

```

see https://github.com/eclipse-platform/eclipse.platform.debug/pull/139#issuecomment-1549217628

**Comment by jukzi** [comment]
This seems to be serious, after only 30 min  of editing Manifests i ran into 
`org.eclipse.swt.SWTError: No more handles`
before i saw a lot of theses stacktraces

**Comment by jukzi** [comment]
> If yes, do you constantly open and close JUnit view?

I did not. i was not able to reproduce yet.  I guess one reason that we normally don't see "SWT Resource was not properly disposed" warnings on images is that they are normally cached with strong references in 
Display.toolDisabledImageList or
LazyResourceManager.map

And only if that caches overflows (which normally does not happen) a warning would be produced.

**Comment by jukzi** [comment]
This is screenshot of a heapdump after working for some hours:
Total images:
![image](https://github.com/eclipse-platform/eclipse.platform/assets/51790620/b7eeb3ff-8b10-4510-976e-f63c0d2fbc1d)
LazyResourceManager:
![image](https://github.com/eclipse-platform/eclipse.platform/assets/51790620/25c13db7-3857-4fbc-a91c-4b1057d1228e)
Display:
![image](https://github.com/eclipse-platform/eclipse.platform/assets/51790620/ed917dd3-cba1-4c11-a729-35ae03a5c0bb)

**Comment by jukzi** [comment]
> Are all the leaked icons referring to `TestRunnerViewPart.createPartControl()` code?

no. also (these errors are only shown when eclipse IDE is closed!):
,,,
	at org.eclipse.jface.action.ActionContributionItem.fill(ActionContributionItem.java:292)
	at org.eclipse.ui.internal.console.OpenConsoleAction.addActionToMenu(OpenConsoleAction.java:125)
or
,,,
	at org.eclipse.jface.action.ActionContributionItem.fill(ActionContributionItem.java:350)
	at org.eclipse.jface.action.ToolBarManager.update(ToolBarManager.java:358)
	at org.eclipse.ui.internal.e4.compatibility.ActionBars.updateActionBars(ActionBars.java:68)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.configureToolBar(TestRunnerViewPart.java:2130)
	at org.eclipse.jdt.internal.junit.ui.TestRunnerViewPart.createPartControl(TestRunnerViewPart.java:1894)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityPart.createPartControl(CompatibilityPart.java:158)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityView.createPartControl(CompatibilityView.java:155)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityPart.create(CompatibilityPart.java:365)


or
...
	at org.eclipse.jface.action.MenuManager.fill(MenuManager.java:277)
	at org.eclipse.jface.action.MenuManager.doItemFill(MenuManager.java:729)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:806)
	at org.eclipse.jface.action.MenuManager.update(MenuManager.java:673)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.scheduleManagerUpdate(MenuManagerRenderer.java:1227)
	at org.eclipse.e4.ui.workbench.renderers.swt.MenuManagerRenderer.processContents(MenuManagerRenderer.java:655)
	at org.eclipse.ui.internal.PopupMenuExtender.addMenuContributions(PopupMenuExtender.java:382)
	at org.eclipse.ui.internal.PopupMenuExtender.menuAboutToShow(PopupMenuExtender.java:355)
	at org.eclipse.jface.action.MenuManager.fireAboutToShow(MenuManager.java:338)
	at org.eclipse.jface.action.MenuManager.handleAboutToShow(MenuManager.java:468)
	at org.eclipse.jface.action.MenuManager$2.menuShown(MenuManager.java:495)


or
....

	at org.eclipse.jface.resource.AbstractResourceManager.create(AbstractResourceManager.java:92)
	at org.eclipse.jface.resource.LazyResourceManager.create(LazyResourceManager.java:89)
	at org.eclipse.jface.resource.LocalResourceManager.allocate(LocalResourceManager.java:71)
	at org.eclipse.jface.resource.AbstractResourceManager.create(AbstractResourceManager.java:92)
	at org.eclipse.jface.resource.ResourceManager.get(ResourceManager.java:145)
	at org.eclipse.egit.ui.internal.UIIcons.getImage(UIIcons.java:569)
	at org.eclipse.egit.ui.internal.actions.SwitchToMenu.<init>(SwitchToMenu.java:108)
	at org.eclipse.egit.ui.internal.actions.SwitchToMenu.<init>(SwitchToMenu.java:95)

**Comment by fedejeanne** [comment]
Hello all, I saw the "help wanted" tag and thought maybe I could look into it.

Is this behavior still reproducible? If yes, any hints on how to do reproduce it?
I read that working for a while with Manifests might do the trick but I haven't been able to reproduce it yet.

If I understand this correctly @jukzi, you saw all those stack-traces in the error log **while running Eclipse**, right? Or did the errors first happen when you closed Eclipse?

**Comment by jukzi** [comment]
Still happens, but i have no recipe to reproduce. My last understanding is that those errors are typically generated in the moment when eclipse is closed after working a lot with it. They are then visible after next eclipse start.

**Comment by fedejeanne** [comment]
> My last understanding is that those errors are typically generated in the moment when eclipse is closed...

Makes sense. I just noticed that the logging happens in `ResourceTracker::finalize()` i.e. when the resource is eligible for GC. By the way relying in the `finalize` method might not be the best way to report such leaks since there is no guarantee that `finalize` will even be called (see the [deprecation notice](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#finalize())) which could leak to unreported leaks.

This might be an overkill for this particular case but seeing that some resources are being forgotten and that they leak, wouldn't it make sense to try and dispose each resource using [`Cleaner`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ref/Cleaner.html) / [`PhantomReference`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/ref/PhantomReference.html) ? That would probably simplify some classes and take away the responsibility of cleaning up from the classes that create the resources. 

What do you think?

**Comment by fedejeanne** [comment]
> This seems to be serious, after only 30 min of editing Manifests i ran into `org.eclipse.swt.SWTError: No more handles` before i saw a lot of theses stacktraces

Do you have a log-file you can post here?

Are you using Linux? If yes, can you please check the maximum of file handles that your OS allows? 

I took some Heap dumps after working with Eclipse for a few hours and noticed that there are up to 1166 instances of `Image` in it. But that number didn't just always go up, it also went down which means that some images are being disposed. 

I will continue to monitor my Eclipse instance and report back if I find something of interest. In the meantime, if you have any new insights, please post them here.

**Comment by jukzi** [comment]
> Do you have a log-file you can post here?

no. my strategy will be: if i get an "No more handles" again i will create a heapdump. I am using windows with a default max of handles. which is "10K per process, and has a total limit of 32K across the system." https://stackoverflow.com/questions/951419/how-many-windows-handles-in-use-is-too-many

**Comment by fedejeanne** [comment]
>  I am using windows with a default max of handles. which is "10K per process, and has a total limit of 32K across the system."

👍 

I have good news and bad news. The good news are I know how to reproduce this error. The bad news are it's a ticking bomb

## How to reproduce it
* Start VisualVM
* Start a **Memory Sampling** for Eclipse
* Filter for **org.eclipse.swt.graphics.Image**
* Go to Eclipse and open any "big" class (e.g. `org.eclipse.jface.action.StatusLine`)
* Search (`Ctrl` + `h`) for the letter "a" in the class and **mark all occurrences**
* **Slowly** scroll down until you reach the end of the class
* **Slowly** scroll up until you reach the beginning of the class
* Repeat the scrolling in the same class

I tried it and I saw how the **live objects** count went up and up, even though it's the same editor and it displays "the same" images again and again. 
![image](https://github.com/eclipse-platform/eclipse.platform/assets/2205684/a78f1e46-b786-4669-ac2a-253cd3f8e50d)

**Comment by jukzi** [comment]
i am sure i never used "mark all occurrences" - so may be a related but different flaw

**Comment by fedejeanne** [comment]
I only marked the occurrences in order to have even more images on the left side of the editor, but the behavior is the same when you have any image. The only difference is that the _live objects_ (i.e. the `Images`) will go up slower.

If you do it without marking occurrences, every time a small image comes into the visible range again, it will generate an instance of `Image`. For example this small arrow or the _minus_ sign

![image](https://github.com/eclipse-platform/eclipse.platform/assets/2205684/fce20362-cef6-4675-bc8e-ff7fa9aa92ca)

**Comment by jukzi** [comment]
ok, sounds broken - should be always the same image.

**Comment by fedejeanne** [comment]
> ok, sounds broken - should be always the same image.

Agreed. That means that the resource "leak" is only a consequence of a resource **exhaustion**. The real problem seems to be that those images are not being reused. Instead, they are being generated over and over and they end up exhausting the file handles, which causes an Exception to be thrown and interrupt the normal LC of a view (the creation/loading). The moment that view becomes `null` all its images become eligible for garbage collection and they are identified as a "leak".

**Comment by fedejeanne** [comment]
To be clear: I see 2 problems here:
1. Resource exhaustion (the images should be reused in order to avoid that)
2. Abnormal termination of a View (or any owner of a `Resource` for that matter). If this happen then the owner still needs to clean up and release all of its resources.

I'll see if I can do anything to address these 2 issues. If you have any hint on where to look first, I would appreciate it. Otherwise I'm gonna' start by looking in the `ResourceManager` & Co. and work my way from there.

Cheers and thank you for your input! :)

**Comment by jukzi** [comment]
but that "live Objects" in visualvm seems to be wrong for this measurement. those images are only alife because GC+finalizer did not run yet. in a heapdump those images are shown as "no GC root"  and do disappear as soon as you press "Run garbage Collector" in eclipse.
there is no referrer to those images other then the ResourceTracker:
![image](https://github.com/eclipse-platform/eclipse.platform/assets/51790620/f4f4db7a-3526-498e-baf2-cb8cf7856722)

**Comment by laeubi** [comment]
> but that "live Objects" in visualvm seems to be wrong for this measurement. those images are only alife because GC+finalizer did not run yet.

The message is issued only if there are no more references and the GC is about to collect so you can't really debug/see this in a heapdump, one has to to investigate the callstack and check why one is releasing the ultimate reference to the object without disposing it.

One reason *might* be that `Image` has a completely crazy equals method defined and one thinks items are equal, or images are stored inside a `Label` or alike but no dispose listener is attached, or in the simplest case an image is created in a local scope and not disposed when leaving the scope.

**Comment by HeikoKlare** [comment]
Even though a potential resource leak is a severe issue, it might make sense to collect some further debug data (i.e., a heapdump when the error occurs) before proceeding with the analysis, because I see at least two "particularities" in SWT error handling which may also lead to a different root cause:
1. From what I've seen so far, SWT's `ERROR_NO_HANDLES` is not only used for actual out-of-handles errors. I faced that at some points in the Edge browser implementation and there also seem to be examples in the `Image` class, like [here](https://github.com/eclipse-platform/eclipse.platform.swt/blob/0cc66f1ea3ffd9263d9c8d0603f0911bce01acee/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/graphics/Image.java#L901) and [here](https://github.com/eclipse-platform/eclipse.platform.swt/blob/0cc66f1ea3ffd9263d9c8d0603f0911bce01acee/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/graphics/Image.java#L1114), where the error actually indicates a failed memory allocation rather than a handle allocation. So the error message might be misleading as something else may have failed and a sufficient number of handles remains.
2. The "SWT Resource was not properly disposed" message will always occur when an SWT error is thrown during initialization. In my opinion, this can be misleading, as it is not a matter of improper disposal but of improper creation. So it might be reasonable if an error during initialization would also clean up the `ResourceTracker` or set the `disposed` state so that it does not print these potentially misleading messages.

**Comment by jukzi** [comment]
> Even though I did triggered a GC during my test several times and the _live objects_ didn't go down

It feels like GC does not include finalization. Objects are collected but not finalized yet.

**Comment by fedejeanne** [comment]
> It feels like GC does not include finalization. Objects are collected but not finalized yet.

Correct, GC != finalization. See this extract from this [deprecation comment](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#finalize())

**Comment by fedejeanne** [comment]
> 1. From what I've seen so far, SWT's ERROR_NO_HANDLES is not only used for actual out-of-handles errors...
> 2. The "SWT Resource was not properly disposed" message will always occur when an SWT error is thrown during initialization...

That would explain why is it so difficult to reproduce by trying to exceed the number of file handles, maybe we're looking in the wrong place and the problem has nothing to do with handles but with some other initialization/heap allocation problem.

BTW I tried to reproduce this in Linux by setting the max amount of file handles to 560 and I couldn't: Linux doesn't give Eclipse more handles beyond this threshold and Eclipse works anyway. For completeness, here's how I did it (I'm using Ubuntu 20.04):

```bash
# set the limit for <PID>
$> sudo prlimit --pid <PID> --nofile=560:560

# check the current nr of file handles for <PID>
$> lsof -a -p <PID> | wc -l
560
```

I won't try to reproduce by playing with the file handles anymore since it's too complicated and it doesn't seem to be the real source of the problem.

I'll look into initialization errors and report back in case I find something.

Thank you all for your input and please post any extra information/logs/dumps you can get your hands on!

Cheers

**Comment by SyntevoAlex** [comment]
I struggled with Eclipse and eventually "reproduced" the invalid problem described in https://github.com/eclipse-platform/eclipse.platform.swt/pull/769

I added this code to the end of `initNonDisposeTracking()`
```java
boolean failAllocation = false;
if (System.currentTimeMillis() - ticks > 60000) {
	if (Math.random() > 0.90)
		failAllocation = true;
}

if (failAllocation) {
	tracker.testException = true;
	SWT.error(SWT.ERROR_UNSPECIFIED);
}
```

When leak is reported, the reported object does NOT have the `testException` flag in it.
This basically means that throwing exception wrecks other Eclipse code that in turn causes a real leak.

TLDR: `ResourceTracker` does not have demonstrated false-positives so far.

**Comment by HeikoKlare** [comment]
> Because before throwing the error, `ResourceTracker` checks for `Resource.isDisposed()`:

> This method, despite its name, usually checks if handle is still allocated.

You are right. I expected the dispose state to be managed explicitly and to be set by calling `dispose()`, but it is obviously identified by the `handle` value. Although the behavior is not consistent across all `Resource` subclasses (e.g., `Color.isDisposed()` will return `false` when initialization is interrupted due to an exception), it seems to be consistent across all classes that might leak resources (`Image`, `Font` etc.).
So I agree with your conclusion that probably some other exception occurs during resource initialization that leaves the resource in an uninitialized state while not being disposed, i.e., while still having a handle assigned.

**Comment by fedejeanne** [comment]
> TLDR: ResourceTracker does not have demonstrated false-positives so far.
You're right, I tried your changes for a while and I haven't seen any occurrences of `tracker.testException == true` so the hypothesis of https://github.com/eclipse-platform/eclipse.platform.swt/pull/769 isn't demonstrated. 

My mistake, thank you for testing it.

**Comment by SyntevoAlex** [comment]
No worries!

**Comment by jukzi** [comment]
closing as not planned.
